### ISSUE TITLE
Terminal does not render true color 

### ISSUE THREAD

**Issue by Tyriar** [xtermjs/xterm.js#484]
See https://github.com/Microsoft/vscode/issues/18544

**Comment by cnsumner** [comment]
Can't wait for this. I've been hacking away at hterm trying to get truecolor support but I feel like this is a way better option.

**Comment by Tyriar** [comment]
Implementing this will need to be done here https://github.com/sourcelair/xterm.js/blob/365a9f949cc1acfb6c4649ee1d85da3bbc60f928/src/InputHandler.ts#L1276

The tricky part is determining how we store the colors against the character(s) and then how to render it (probably inline `style` attributes). Something like https://github.com/sourcelair/xterm.js/pull/450 would probably simplify the adding of the attributes.

**Comment by cnsumner** [comment]
I may take a look at it when I get some time, but I can't say I'm too familiar with control sequences.

**Comment by jerch** [comment]
@cnsumner xterm currently cant display true color but has support for one of the proposed escape sequences. You can read more about this here: https://gist.github.com/XVilka/8346728

**Comment by Tyriar** [comment]
Put together a messy proof of concept to get a better idea of what we need to do here https://github.com/Tyriar/xterm.js/tree/484_truecolor

![image](https://user-images.githubusercontent.com/2193314/27773235-3230d816-5f2a-11e7-9850-2420a43f1cbd.png)

Things that need to be done:

- The branch currently adds another number to each character representing it's 'truecolor' in the format `0x1RRGGBB` (fg) or `0x0RRGGBB` (bg), which fits in a 32 bit integer
- It needs to support *both* fg and bg colors, and be able to mix truecolor and regular color
- `curTrueColor` is super ugly
- Support inverse
- It might be time to rethink how characters are encoded, the current format is:

  ```
  [attribute, character, width]
  ```

  - `attribute`: A 32-bit integer that holds ascii fg, ascii bg and flags (bold, underline, blink, inverse, invisible), not all bits are used
  - `char`: Either the empty string (for 0 width chars that pad wide chars) or a single character unicode string
  - `width`: The width of the char, currently can be 0, 1 or 2 but this could include more to support tabs properly https://github.com/sourcelair/xterm.js/issues/734
- The eventual solution should be compact for memory and also quick to access, sticking with an array, binary flags and bit shifting for data is probably the best idea.

**Comment by jerch** [comment]
@Tyriar 
Yeah memory usage will explode if it is not optimized for low consumption. Maybe it can be packed together with `width`, this field is unlikely to get any bigger than 16 consuming only 4 bytes (tab defaults to 4 or 8 normally). This leaves room for one RGB definition* (assuming `width` is a 32bit integer). Not sure how many bits are free in `attribute`, maybe there can go the second RGB definition*. The 256 color definitions could end up in one of those since they can't be defined simultaneously with RGB on one cell.
(*) A cell will need room for two RGB definitions for foreground and background.

**Comment by Tyriar** [comment]
It looks like attributes currently consumed 9 bytes for bg, 9 bytes for bg and 5 bytes for flags, leaving plenty to squeeze width in at the end. The RGB definitions need 24 bytes (8*3) each.

Here's the (slightly complicated) space saving approach I think you're suggesting:

```
[char, attr1, attr2]
```

Where attr1 is (29 bits):

- width: 4 bits
- is 256 color or truecolor flag: 1 bit
- bg color: 24 bits

And attr2 is (30 bits):

- flags: 5 bits
- is 256 color or truecolor flag: 1 bit
- fg color: 24 bits

---

A simpler approach would use a number specifically for attributes:

```
[char, attr, truecolorBg, truecolorFg]
```

Where attr is:

- width: 4 bytes
- flags: 5 bytes
- 256 bg color: 9 bytes
- 256 fg color: 9 bytes

---

I want so badly to encapsulate these details as they really complicate things, I wonder if typescript will let us put getters on an array somehow. Using an object was 20x slower than an array based on a micro benchmark I did on jsperf, I know they're not always reliable but this is kind of what I expected as overhead to create an object based on a prototype.

Whatever we do, the format needs to be documented very well. Right now you have to figure it out yourself by looking at `InputHandler.charAttributes`.

**Comment by Tyriar** [comment]
Some more thoughts on memory saving:

- If a character has a width of > 1, we don't really need to store any data at all for the following characters. The "empty characters" are there as a workaround and would probably work fine being `null`.
- Normally a set of characters is styled, meaning attributes could be shared between a sequence of characters.
- Attributes could be cached and each char uses a pointer to one of the cached values,

**Comment by jerch** [comment]
Yes I think on most JS engines arrays with index access are still faster than objects and their property resolving. V8 got some extra optimizations for objects though. Construction and GC wise an array is faster due to the much simpler memory layout (this disadvantage can be omitted by reutilizing existing objects).

I have no clue about the right layout here, problem is that any higly packed stuff will raise runtime penalty due to the needed transformations. (This is one reason not to use bitfields in C if memory is no concern - the needed shift operations will pollute the instruction cache leading to cache misses - the code runs much slower).
Imho only tests will show what works best here.

**Comment by Tyriar** [comment]
@jerch interesting, I have 4 number version for simplicity right now in https://github.com/sourcelair/xterm.js/pull/756.

So memory here is expected to be around: 5 * (1000 scrollback) * (8 bytes on 64-bit machine) * 80 = 3.2mb (before overhead of js array). The actual width of a typical VS Code terminal would see that double if we keep the current format. Seems pretty steep now that I look at it, especially since users will often have higher than 1000 scrollback and multiple active terminals. The trailing whitespace which really doesn't add much at all and could just as easily be null which would probably be a pretty huge win.

The attribute/color set cache idea also seems promising, as long as there's a good way for fast access. A lightweight splay tree-like structure would be ideal (fast access for attributes uses most frequently).

It might also be worth looking into reusing line data arrays when they're trimmed instead of just discarding them ([eg. here](https://github.com/sourcelair/xterm.js/blob/master/src/utils/CircularList.ts#L133)).

**Comment by jerch** [comment]
Yeah and with the engine overhead in mind it grows even bigger - everything in an array or object is stored as a reference, basically a pointer to the content. Dense js-arrays have a slight advantage here, they are formed as a C like array somewhere down the path. Sparse arrays and others objects are build with some tree magic to solve the property resolution.
So for 64bit the reference thing adds 5 * 8 more bytes per cell (it basically doubles your calculation above - 6.4mb). The content can add even more bytes beside the "real content", but this depends on the actual implementation of the Integer and String type.

Maybe a TypedArray can solve this. There the integer stuff can reside directly at the index position without an additional indirection. asm.js basically operates only on those.

**Comment by jerch** [comment]
Here is a quick micro benchmark to compare TypedArray and js-Array for creation and data insertion: http://jsbench.github.io/#af3ea1056a70df2aa6775699d174df0d

The TypedArray runs 10 times faster with FF and Chrome on my machine. This might be due to the preallocation, have not tested this with js-Array since it is discouraged. Not sure if the JITs spot the fact, that the value just increases by 1 every 8 bytes and even optimize this out. Then the benchmark is useless at all.

The access is slightly worse for a TypedArray (maybe due to `Number` conversion), memory consumption is much better with a TypedArray.

**Comment by szmarczak** [comment]
Hello. Any updates on this? Can't wait to get this working 😃

**Comment by vastbinderj** [comment]
Is anyone actively working this issue?  Is it being discussed anywhere else? @chabou @ty? Is Microsoft helping given the usage in VS Code?

This is the last lagging issue for using hyper in WSL on Windows 10 to provide a decent dev CLI workflow supporting tmux and vim/neovim. Other terminal emulators available to use with WSL just aren't pleasant to look at, wsltty is close, but tab support and plugin support is missing...

**Comment by rfgamaral** [comment]
@vastbinderj Why the missing true color support doesn't give you a decent dev CLI workflow? Honest question.

**Comment by jerch** [comment]
@szmarczak and @vastbinderj 
This issue is still on the agenda, but kinda blocked due to the high memory usage it will introduce with the current terminal buffer design. It will get more attention when #791 has brought us a better buffer layout.

**Comment by vastbinderj** [comment]
@rfgamaral simple - I work across many many different unix/linux servers a day on very large enterprise systems with many microservices and not having decent color support in the main host is disruptive. To deal with the colors not being the same whether I'm on rhel or centos or debian, is painful. Also, sitting down to help a jr or peer developer on windows means I have to context switch if things don't work uniformly.  However, I am super impressed how far wsl has come so quickly with improvements released on an almost monthly basis and the same with VS Code, it is outstanding. 

@jerch Thanks for the update, will wait patiently for #791 to be worked out...

**Comment by jcjolley** [comment]
Greatly looking forward to this as well.

**Comment by jerch** [comment]
@PerBothner We decided to go with a Uint32Array for the new buffer, and for convenience to put the colors into that array as well. This is not done yet as we want to sanitize the new impl first before supporting truecolor.

Also I did some tests/calculations about possible layouts (see https://gist.github.com/jerch/27c2cf91ad313a25415873e4047b2900). Long story short - the ideas are about the tradeoff mem saving vs. runtime penalty. The transition to typed array already gave huge mem saving (currently implements the second idea from above, though the true color values are not applied yet).
